tamr_toolbox/utils/downstream.py

```diff
@@ -0,0 +1,245 @@
+            dependencies suggested by name matching.
+        all_tamr_datasets: A dictionary of all datasets existing in Tamr by name. Default to None.
+        all_unified_datasets: A List of current unified datasets in Tamr. Default to None
+        downstream_datasets: A dictionary of downstream datasets by name. Default to None.
+
+    Returns:
+        List of Dataset objects.
+
+    """
+    # initialize optional variables
+    if not all_tamr_datasets:
+        all_tamr_datasets = _get_all_tamr_datasets(client)
+    if not all_unified_datasets:
+        all_unified_datasets = _get_all_unified_datasets(client)
+    if not downstream_datasets:
+        downstream_datasets = {}
+
+    # Initialize a set of datasets need to check further downstream
+    # (if include dependencies by name)
+    datasets_check = set()
+    LOGGER.debug(f"Start checking dependencies of {dataset_name}")
+    dataset = all_tamr_datasets[dataset_name]
+    dataset_dependencies = dataset.usage().dependencies
+    LOGGER.debug(
+        f"Dataset {dataset.name} has {len(dataset_dependencies)} "
+        f"datasets with dependency suggested by pipeline"
+    )
+    for d in dataset_dependencies:
+        if include_dependencies_by_name and _is_unified_dataset(
+            d.dataset_name, all_unified_datasets=all_unified_datasets
+        ):
+            # if a downstream dataset is unified dataset and want to include dependencies by name,
+            # need to check it further downstream
+            datasets_check.add(d.dataset_name)
+        elif d.dataset_name not in downstream_datasets.keys():
+            downstream_datasets[d.dataset_name] = d.dataset()
+
+    # get dependencies suggested by name matching
+    # if the target dataset is or used to be unified dataset
+    # and whose prefix is the target dataset's name if the target is unified dataset
+    if include_dependencies_by_name and _is_unified_dataset(
+        dataset_name, all_unified_datasets=all_unified_datasets
+    ):
+        dependencies_suggested_by_name = set(
+            [d for d in all_tamr_datasets.keys() if d.startswith(dataset_name)]
+        )
+        LOGGER.debug(
+            f"Dataset {dataset_name} has {len(dependencies_suggested_by_name)} "
+            f"datasets with dependency suggested by name"
+        )
+        datasets_check = datasets_check.union(dependencies_suggested_by_name)
+        # Filter out datasets that have already been checked and the target dataset
+        datasets_check = datasets_check.difference(downstream_datasets.keys()).difference(
+            {dataset.name}
+        )
+
+    # add the target dataset to list of downstream datasets to avoid checking later
+    downstream_datasets[dataset_name] = dataset
+
+    # check further downstream for either dependencies suggested by name
+    # or dependencies that used to or is unified dataset
+    for d in datasets_check:
+        _find_downstream_datasets(
```

Comment:
Nice! Can you just add a comment saying that it relies on passing `downstream_datasets` by reference to update the dictionary?

It may be a bit confusing to someone reading since you also return downstream_datasets explicitly. I think this is fine because the user is not supposed to pass the `downstream_datasets` object directly when using it. For that reason might also be useful to make a comment in the docstring that this and other arguments are not to be set by the user and only used internally

Reply:
Thanks! I added the comment. This is an internal function so do you think we should still add the disclaimer in docs string?

Reply:
Yes, couldn't hurt

Reply:
added to docs string, thanks Alex!